chall10/edramire.py

```diff
@@ -0,0 +1,110 @@
+#! /usr/bin/env python3
+
+import sys
+import re
+from enum import Enum
+
+
+class Cell(Enum):
+    EMPTY = '.'
+    RED = 'O'
+    BLUE = 'X'
+
+    def __str__(self):
+        return str(self.value)
+
+    def __repr__(self):
+        return self.__str__()
+
+
+class Connect:
+    def __init__(self):
+        self.grid = [[Cell.EMPTY for _ in range(7)] for _ in range(6)]
+        self.blue = None
+        self.red = None
+        self.win = None
+
+    def down(self, row: int, col: int, cell: Cell):
+        count = 0
+        while row < 6 and count < 4:
+            if cell == self.grid[row][col]:
+                count += 1
+            else:
+                break
+            row += 1
+        return count == 4
+
+    def check_line(self, m: int, b: int, cell: Cell):
+        count = 0
+        col = 0
+        while col < 7 and count < 4:
+            row = m * col + b
+            if 0 <= row < 6:
+                if cell == self.grid[row][col]:
+                    count += 1
+                elif count < 4:
+                    count = 0
+            col += 1
+        return count == 4
+
+    def load_line(self, line: str, cell):
+        if re.search('^[A-G]_[A-Z][a-z]+$', line) is not None:
```

Comment:
Nice regex ;), only fail I see is that `A_Team2` or `A_TEAM` should be valid too !